modules/sdk-coin-sol/src/sol.ts

```diff
@@ -1043,15 +1041,63 @@ export class Sol extends BaseCoin {
     }
 
     const factory = this.getBuilder();
-    const walletCoin = this.getChain();
 
     let txBuilder;
-    const blockhash = await this.getBlockhash();
+    let blockhash;
+    const recovertTxns: BaseBroadcastTransactionResult[] = [];
 
     const rentExemptAmount = await this.getRentExemptAmount();
 
-    // if this is closeATA recovery
+    // do token recovery before closing ATA address
+    // check if any token is present on the closeAtaAddress
+    const tokenInfo = await this.getTokenAccountInfo(params.closeAtaAddress);
+    const tokenBalance = Number(tokenInfo.info.tokenAmount.amount);
+
+    if (tokenBalance > 0) {
+      // closeATA address has some token balance, it needs to be withdrawn before closing ATA
+      console.log(
+        `closeATA address ${params.closeAtaAddress} has token balance ${tokenBalance}, it needs to be withdrawn before closing ATA address`
+      );
+
+      if (!params.recoveryDestinationAtaAddress || !this.isValidAddress(params.recoveryDestinationAtaAddress)) {
+        throw new Error('invalid recoveryDestinationAtaAddress');
+      }
+
+      blockhash = await this.getBlockhash();
+
+      txBuilder = factory
+        .getTokenTransferBuilder()
+        .nonce(blockhash)
+        .sender(bs58EncodedPublicKey)
+        .fee({ amount: feePerSignature })
+        .associatedTokenAccountRent(rentExemptAmount.toString())
+        .feePayer(bs58EncodedPublicKey);
+
+      const network = this.getNetwork();
+      const token = getSolTokenFromAddress(tokenInfo.info.mint, network);
```

Comment:
what if there are multiple tokens ?

Reply:
There won't be multiple tokens, we are not dealing with receive address here, but we are only interested in closeATA address, and a ATA address will always belong to just one token.